core/src/lib.rs

```diff
@@ -48,6 +48,9 @@ pub type RolesMap = HashMap<RoleId, Role>;
 /// API to work with a collections of [`AccountId`] [`Permissions`] mappings.
 pub type PermissionTokensMap = HashMap<AccountId, Permissions>;
 
+/// API to work with a collections of [`AccountId`][`RoleId`] mappings.
+pub type AccountRolesSet = BTreeSet<role::RoleIdWithOwner>;
```

Comment:
Do you consider to use `BTreeMap<AccountId, BTreeSet<RoleId>>`? Could you please explain why using pairs (`RoleIdWithOwner`s) is better? I assume it has the same time complexity and doesn't require implementing `MaxMinExt`.

Reply:
Yes, i've considered this option.
My primary motivation was to reduce amount of pointer indirection and thus optimize for cases where we scan through all accounts and all roles in it (e.g. when role is removed we have to remove it from every account).

Reply:
Also i believe `MaxMinExt` might be useful in different context as well (if we will decide to implement #3921).

Reply:
But the downside of this approach is memory footprint. I think we should start considering string interning (or string pooling)

Reply:
> But the downside of this approach is the memory footprint. I think we should start considering string interning (or string pooling)

Yeah, about string pooling, if `domain` and `account` are small enough (less 16 bytes each) they will be inlined and no allocation will happen.